addons/sale_stock/tests/test_sale_stock.py

```diff
@@ -734,3 +734,64 @@ def test_12_return_without_refund(self):
         return_picking.button_validate()
         # Checks the delivery amount (must still be 10).
         self.assertEqual(sale_order.order_line.qty_delivered, 10)
+
+    def test_08_sale_return_qty_and_cancel(self):
+        """
+        Test a SO with a product on delivery. Deliver the SO,
+        then do a return of the delivered picking (for all quantities).
+        Then cancel Sale order, it won't raise any warning, it should be cancelled.
+        """
+        partner = self.env['res.partner'].create({
+            'name': 'Marc',
+            'email': 'marc@task.com',
+        })
+        product = self.env['product.product'].create({
+            'name': 'Office Chair',
+            'categ_id': self.env.ref('product.product_category_all').id,
+            'standard_price': 50.0,
+            'list_price': 70.0,
+            'type': 'consu',
+            'uom_id': self.env.ref('uom.product_uom_unit').id,
+            'uom_po_id': self.env.ref('uom.product_uom_unit').id,
+        })
+        so_vals = {
+            'partner_id': partner.id,
+            'partner_invoice_id': partner.id,
+            'partner_shipping_id': partner.id,
+            'order_line': [(0, 0, {
+                'name': product.name,
+                'product_id': product.id,
+                'product_uom_qty': 5.0,
+                'product_uom': product.uom_id.id,
+                'price_unit': product.list_price})],
+            'pricelist_id': self.env.ref('product.list0').id,
+        }
+        so = self.env['sale.order'].create(so_vals)
+
+        # confirm our standard so, check the picking
+        so.action_confirm()
+        self.assertTrue(so.picking_ids, 'Sale Stock: no picking created for "invoice on delivery" storable products')
+
+        # deliver completely
+        pick = so.picking_ids
+        pick.move_lines.write({'quantity_done': 5})
+        pick.button_validate()
+
+        # Create return picking
+        stock_return_picking_form = Form(self.env['stock.return.picking'].with_context(active_ids=pick.ids, active_id=pick.ids[0], active_model='stock.picking'))
+        return_wiz = stock_return_picking_form.save()
+        return_wiz.product_return_moves.quantity = 5.0
+        return_wiz.product_return_moves.to_refund = True
+        res = return_wiz.create_returns()
+        return_pick = self.env['stock.picking'].browse(res['res_id'])
+
+        # Validate picking
+        return_pick.move_lines.write({'quantity_done': 5})
+        return_pick.button_validate()
+
+        # Cancel sale order
+        so.action_cancel()
+        wizard = self.env['sale.order.cancel'].with_context({'order_id': so.id}).create({'order_id': so.id})
+        wizard.action_cancel()
```

Comment:
assert that the draft invoice is cancelled and that the posted one is still posted